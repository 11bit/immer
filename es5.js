"use strict"
// @ts-check

const PROXY_TARGET = Symbol("immer-proxy")
const CHANGED_STATE = Symbol("immer-changed-state")
const PARENT = Symbol("immer-parent")

let autoFreeze = true

/**
 * Immer takes a state, and runs a function against it.
 * That function can freely mutate the state, as it will create copies-on-write.
 * This means that the original state will stay unchanged, and once the function finishes, the modified state is returned
 *
 * @export
 * @param {any} baseState - the state to start with
 * @param {Function} thunk - function that receives a proxy of the base state as first argument and which can be freely modified
 * @returns {any} a new state, or the base state if nothing was modified
 */
function immer(baseState, thunk) {
    let finalizing = false
    let finished = false
    const descriptors = {}
    const proxies = []

    // creates a proxy for plain objects / arrays
    function createProxy(base, parent) {
        let proxy
        if (isPlainObject(base)) proxy = createObjectProxy(base)
        else if (Array.isArray(base)) proxy = createArrayProxy(base)
        else throw new Error("Expected a plain object or array")
        createHiddenProperty(proxy, PROXY_TARGET, base)
        createHiddenProperty(proxy, CHANGED_STATE, false)
        createHiddenProperty(proxy, PARENT, parent)
        proxies.push(proxy)
        return proxy
    }

    function assertUnfinished() {
        if (finished)
            throw new Error(
                "Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process?"
            )
    }

    function proxySet(proxy, prop, value) {
        // immer func not ended?
        assertUnfinished()
        // actually a change?
        if (Object.is(proxy[prop], value)) return
        // mark changed
        markDirty(proxy)
        // and stop proxying, we know this object has changed
        Object.defineProperty(proxy, prop, {
            enumerable: true,
            writable: true,
            configurable: true,
            value: value
        })
    }

    function createPropertyProxy(prop) {
        return (
            descriptors[prop] ||
            (descriptors[prop] = {
                configurable: true,
                enumerable: true,
                get() {
                    assertUnfinished()
                    // find the target object
                    const target = this[PROXY_TARGET]
                    // find the original value
                    const value = target[prop]
                    // if we are finalizing, don't bother creating proxies, just return base value
                    if (finalizing) return value
                    // if not proxy-able, return value
                    if (!isPlainObject(value) && !Array.isArray(value))
                        return value
                    // otherwise, create proxy
                    const proxy = createProxy(value, this)
                    // and make sure this proxy is returned from this prop in the future if read
                    // (write behavior as is)
                    Object.defineProperty(this, prop, {
                        configurable: true,
                        enumerable: true,
                        get() {
                            return proxy
                        },
                        set(value) {
                            proxySet(this, prop, value)
                        }
                    })
                    return proxy
                },
                set(value) {
                    proxySet(this, prop, value)
                }
            })
        )
    }

    function createObjectProxy(base) {
        const proxy = Object.assign({}, base)
        Object.keys(base).forEach(prop =>
            Object.defineProperty(proxy, prop, createPropertyProxy(prop))
        )
        return proxy
    }

    function createArrayProxy(base) {
        const proxy = new Array(base.length)
        for (let i = 0; i < base.length; i++)
            Object.defineProperty(proxy, "" + i, createPropertyProxy("" + i))
        return proxy
    }

    // this sounds very expensive, but actually it is not that extensive in practice
    // as it will only visit proxies, and only do key-based change detection for objects for
    // which it is not already know that they are changed (that is, only object for which no known key was changed)
    function markChanges() {
        // intentionally we process the proxies in reverse order;
        // ideally we start by processing leafs in the tree, because if a child has changed, we don't have to check the parent anymore
        // reverse order of proxy creation approximates this
        for (let i = proxies.length - 1; i >= 0; i--) {
            const proxy = proxies[i]
            if (
                proxy[CHANGED_STATE] === false &&
                ((isPlainObject(proxy) && hasObjectChanges(proxy)) ||
                    (Array.isArray(proxy) && hasArrayChanges(proxy)))
            ) {
                markDirty(proxy)
            }
        }
    }

    function hasObjectChanges(proxy) {
        const baseKeys = Object.keys(proxy[PROXY_TARGET])
        const keys = Object.keys(proxy)
        return !shallowEqual(baseKeys, keys)
    }

    function hasArrayChanges(proxy) {
        return proxy[PROXY_TARGET].length !== proxy.length
    }

    function finalize(proxy) {
        // given a base object, returns it if unmodified, or return the changed cloned if modified
        if (!isProxy(proxy)) return proxy
        if (!proxy[CHANGED_STATE]) return proxy[PROXY_TARGET] // return the original target
        if (isPlainObject(proxy)) return finalizeObject(proxy)
        if (Array.isArray(proxy)) return finalizeArray(proxy)
        throw new Error("Illegal state")
    }

    function finalizeObject(proxy) {
        const res = {}
        Object.keys(proxy).forEach(prop => {
            res[prop] = finalize(proxy[prop])
        })
        return freeze(res)
    }

    function finalizeArray(proxy) {
        return freeze(proxy.map(finalize))
    }

    // create proxy for root
    const rootClone = createProxy(baseState, undefined)
    // execute the thunk
    const maybeVoidReturn = thunk(rootClone)
    //values either than undefined will trigger warning;
    !Object.is(maybeVoidReturn, undefined) &&
        console.warn(
            `Immer callback expects no return value. However ${typeof maybeVoidReturn} was returned`
        )
    // and finalize the modified proxy
    finalizing = true
    // find and mark all changes (for parts not done yet)
    markChanges()
    const res = finalize(rootClone)
    // make sure all proxies become unusable
    finished = true
    return res
}

function markDirty(proxy) {
    proxy[CHANGED_STATE] = true
    let parent = proxy
    while ((parent = parent[PARENT])) {
        if (parent[CHANGED_STATE] === true) return
        parent[CHANGED_STATE] = true
    }
}

function isPlainObject(value) {
    if (value === null || typeof value !== "object") return false
    const proto = Object.getPrototypeOf(value)
    return proto === Object.prototype || proto === null
}

function isProxy(value) {
    return !!(value && value[PROXY_TARGET])
}

function freeze(value) {
    if (autoFreeze) {
        Object.freeze(value)
    }
    return value
}

function createHiddenProperty(target, prop, value) {
    Object.defineProperty(target, prop, {
        value: value,
        enumerable: false,
        writable: true
    })
}

function shallowEqual(objA, objB) {
    //From: https://github.com/facebook/fbjs/blob/c69904a511b900266935168223063dd8772dfc40/packages/fbjs/src/core/shallowEqual.js
    if (Object.is(objA, objB)) return true
    if (
        typeof objA !== "object" ||
        objA === null ||
        typeof objB !== "object" ||
        objB === null
    ) {
        return false
    }
    const keysA = Object.keys(objA)
    const keysB = Object.keys(objB)
    if (keysA.length !== keysB.length) return false
    for (let i = 0; i < keysA.length; i++) {
        if (
            !hasOwnProperty.call(objB, keysA[i]) ||
            !Object.is(objA[keysA[i]], objB[keysA[i]])
        ) {
            return false
        }
    }
    return true
}

/**
 * Automatically freezes any state trees generated by immer.
 * This protects against accidental modifications of the state tree outside of an immer function.
 * This comes with a performance impact, so it is recommended to disable this option in production.
 * It is by default enabled.
 *
 * @returns {void}
 */
function setAutoFreeze(enableAutoFreeze) {
    autoFreeze = enableAutoFreeze
}

createHiddenProperty(exports, "__esModule", true)
module.exports.default = immer
module.exports.setAutoFreeze = setAutoFreeze
