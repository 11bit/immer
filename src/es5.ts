"use strict"
import {
	each,
	has,
	is,
	isDraft,
	isDraftable,
	isEnumerable,
	isMap,
	isSet,
	shallowCopy,
	DRAFT_STATE,
	latest
} from "./common"

import {ImmerScope} from "./scope"
import {ImmerState, Drafted, AnyObject, AnyMap} from "./types"

// TODO: merge with ImmerState?
interface ES5BaseState {
	scope: ImmerScope
	modified: boolean
	finalizing: boolean
	finalized: boolean
	assigned: {[key: string]: any}
	parent?: ImmerState
	drafts: AnyMap | null // TODO: don't use map?
	revoke()
	revoked: boolean
}

export interface ES5ObjectState extends ES5BaseState {
	type: "es5_object"
	draft: Drafted<AnyObject, ES5ObjectState>
	base: AnyObject
	copy: AnyObject | null
}

export interface ES5ArrayState extends ES5BaseState {
	type: "es5_array"
	draft: Drafted<AnyObject, ES5ArrayState>
	base: AnyObject
	copy: AnyObject | null
}

type ES5State = ES5ArrayState | ES5ObjectState

export function willFinalizeES5(
	scope: ImmerScope,
	result: any,
	isReplaced: boolean
) {
	scope.drafts!.forEach(draft => {
		draft[DRAFT_STATE].finalizing = true
	})
	if (!isReplaced) {
		if (scope.patches) {
			markChangesRecursively(scope.drafts![0])
		}
		// This is faster when we don't care about which attributes changed.
		markChangesSweep(scope.drafts)
	}
	// When a child draft is returned, look for changes.
	else if (isDraft(result) && result[DRAFT_STATE].scope === scope) {
		markChangesSweep(scope.drafts)
	}
}

export function createES5Proxy<T>(
	base: T,
	parent?: ImmerState
): Drafted<any, ES5ObjectState | ES5ArrayState> {
	const isArray = Array.isArray(base)
	const draft = clonePotentialDraft(base)

	each(draft, prop => {
		proxyProperty(draft, prop, isArray || isEnumerable(base, prop))
	})

	const state: ES5ObjectState | ES5ArrayState = {
		type: isArray ? "es5_array" : ("es5_object" as any),
		scope: parent ? parent.scope : ImmerScope.current!,
		modified: false,
		finalizing: false, // es5 only
		finalized: false,
		assigned: {},
		parent,
		base,
		draft,
		drafts: null,
		copy: null,
		revoke,
		revoked: false // es5 only
	}

	createHiddenProperty(draft, DRAFT_STATE, state)
	return draft
}

function revoke(this: ES5BaseState) {
	this.revoked = true
}

// TODO: type these methods
// Access a property without creating an Immer draft.
function peek(draft, prop) {
	const state = draft[DRAFT_STATE]
	if (state && !state.finalizing) {
		state.finalizing = true
		const value = draft[prop]
		state.finalizing = false
		return value
	}
	return draft[prop]
}

function get(state, prop) {
	assertUnrevoked(state)
	const value = peek(latest(state), prop)
	if (state.finalizing) return value
	// Create a draft if the value is unmodified.
	if (value === peek(state.base, prop) && isDraftable(value)) {
		prepareCopy(state)
		return (state.copy[prop] = state.scope.immer.createProxy(value, state))
	}
	return value
}

function set(state: ES5State, prop, value) {
	assertUnrevoked(state)
	state.assigned[prop] = true
	if (!state.modified) {
		if (is(value, peek(latest(state), prop))) return
		markChangedES5(state)
		prepareCopy(state)
	}
	state.copy![prop] = value
}

export function markChangedES5(state) {
	if (!state.modified) {
		state.modified = true
		if (state.parent) markChangedES5(state.parent)
	}
}

function prepareCopy(state) {
	if (!state.copy) state.copy = clonePotentialDraft(state.base)
}

function clonePotentialDraft(base) {
	const state = base && base[DRAFT_STATE]
	if (state) {
		state.finalizing = true
		const draft = shallowCopy(state.draft, true)
		state.finalizing = false
		return draft
	}
	return shallowCopy(base)
}

// property descriptors are recycled to make sure we don't create a get and set closure per property,
// but share them all instead
const descriptors = {}

function proxyProperty(draft, prop, enumerable) {
	let desc = descriptors[prop]
	if (desc) {
		desc.enumerable = enumerable
	} else {
		descriptors[prop] = desc = {
			configurable: true,
			enumerable,
			get() {
				return get(this[DRAFT_STATE], prop)
			},
			set(value) {
				set(this[DRAFT_STATE], prop, value)
			}
		}
	}
	Object.defineProperty(draft, prop, desc)
}

export function assertUnrevoked(state) {
	if (state.revoked === true)
		throw new Error(
			"Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process? " +
				JSON.stringify(latest(state))
		)
}

// This looks expensive, but only proxies are visited, and only objects without known changes are scanned.
function markChangesSweep(drafts) {
	// The natural order of drafts in the `scope` array is based on when they
	// were accessed. By processing drafts in reverse natural order, we have a
	// better chance of processing leaf nodes first. When a leaf node is known to
	// have changed, we can avoid any traversal of its ancestor nodes.
	for (let i = drafts.length - 1; i >= 0; i--) {
		const state = drafts[i][DRAFT_STATE]
		if (!state.modified) {
			if (Array.isArray(state.base)) {
				if (hasArrayChanges(state)) markChangedES5(state)
			} else if (isMap(state.base)) {
				// TODO: use switch
				continue
			} else if (isSet(state.base)) {
				continue
			} else if (hasObjectChanges(state)) {
				markChangedES5(state)
			}
		}
	}
}

// TODO: refactor this to work per object-type
// TODO: Set / Map shouldn't be ES specific
function markChangesRecursively(object) {
	if (!object || typeof object !== "object") return
	const state = object[DRAFT_STATE]
	if (!state) return
	const {base, draft, assigned} = state
	if (isSet(object) || isMap(object)) {
		// TODO: create switch here
		return
	} else if (!Array.isArray(object)) {
		// Look for added keys.
		// TODO: use each?
		Object.keys(draft).forEach(key => {
			// The `undefined` check is a fast path for pre-existing keys.
			if (base[key] === undefined && !has(base, key)) {
				// TODO: this code seems invalid for Maps!
				assigned[key] = true
				markChangedES5(state)
			} else if (!assigned[key]) {
				// TODO: === false ?
				// Only untouched properties trigger recursion.
				markChangesRecursively(draft[key])
			}
		})
		// Look for removed keys.
		// TODO: is this code needed?
		// TODO: use each?
		Object.keys(base).forEach(key => {
			// The `undefined` check is a fast path for pre-existing keys.
			if (draft[key] === undefined && !has(draft, key)) {
				assigned[key] = false
				markChangedES5(state)
			}
		})
	} else if (hasArrayChanges(state)) {
		markChangedES5(state)
		assigned.length = true
		if (draft.length < base.length) {
			for (let i = draft.length; i < base.length; i++) assigned[i] = false
		} else {
			for (let i = base.length; i < draft.length; i++) assigned[i] = true
		}
		for (let i = 0; i < draft.length; i++) {
			// Only untouched indices trigger recursion.
			if (assigned[i] === undefined) markChangesRecursively(draft[i])
		}
	}
}

function hasObjectChanges(state) {
	const {base, draft} = state

	// Search for added keys and changed keys. Start at the back, because
	// non-numeric keys are ordered by time of definition on the object.
	const keys = Object.keys(draft)
	for (let i = keys.length - 1; i >= 0; i--) {
		const key = keys[i]
		const baseValue = base[key]
		// The `undefined` check is a fast path for pre-existing keys.
		if (baseValue === undefined && !has(base, key)) {
			return true
		}
		// Once a base key is deleted, future changes go undetected, because its
		// descriptor is erased. This branch detects any missed changes.
		else {
			const value = draft[key]
			const state = value && value[DRAFT_STATE]
			if (state ? state.base !== baseValue : !is(value, baseValue)) {
				return true
			}
		}
	}

	// At this point, no keys were added or changed.
	// Compare key count to determine if keys were deleted.
	return keys.length !== Object.keys(base).length
}

function hasArrayChanges(state) {
	const {draft} = state
	if (draft.length !== state.base.length) return true
	// See #116
	// If we first shorten the length, our array interceptors will be removed.
	// If after that new items are added, result in the same original length,
	// those last items will have no intercepting property.
	// So if there is no own descriptor on the last position, we know that items were removed and added
	// N.B.: splice, unshift, etc only shift values around, but not prop descriptors, so we only have to check
	// the last one
	const descriptor = Object.getOwnPropertyDescriptor(draft, draft.length - 1)
	// descriptor can be null, but only for newly created sparse arrays, eg. new Array(10)
	if (descriptor && !descriptor.get) return true
	// For all other cases, we don't have to compare, as they would have been picked up by the index setters
	return false
}

function createHiddenProperty(target, prop, value) {
	Object.defineProperty(target, prop, {
		value: value,
		enumerable: false,
		writable: true
	})
}
