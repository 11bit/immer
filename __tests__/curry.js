"use strict"
import * as immerProxy from ".."
import * as immerEs5 from "../es5"

runTests("proxy", immerProxy)
runTests("es5", immerEs5)

function runTests(name, lib) {
    describe("auto freeze - " + name, () => {
        const produce = lib.default

        it("should check arguments", () => {
            expect(() => produce()).toThrow(/produce expects 1 or 2 arguments/)
            expect(() => produce(() => {}, {})).toThrow(
                /the first argument to produce should be a plain object or array/
            )
            expect(() => produce(1, () => {})).toThrow(
                /the first argument to produce should be a plain object or array/
            )
            expect(() => produce({}, () => {}, [])).toThrow(
                /produce expects 1 or 2 arguments/
            )
        })

        it("should support currying", () => {
            const state = [{}, {}, {}]
            const mapper = produce((item, index) => {
                item.index = index
            })

            expect(state.map(mapper)).not.toBe(state)
            expect(state.map(mapper)).toEqual([
                {index: 0},
                {index: 1},
                {index: 2}
            ])
        })
    })
}
