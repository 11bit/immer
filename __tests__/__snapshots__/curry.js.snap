// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`curry - es5 should check arguments 1`] = `"Invariant failed: The first or second argument to \`produce\` must be a function"`;

exports[`curry - es5 should check arguments 2`] = `"Invariant failed: The first or second argument to \`produce\` must be a function"`;

exports[`curry - es5 should check arguments 3`] = `"Invariant failed: The first or second argument to \`produce\` must be a function"`;

exports[`curry - es5 should check arguments 4`] = `"Invariant failed: The third argument to \`produce\` must be a function or undefined"`;

exports[`curry - proxy should check arguments 1`] = `"Invariant failed: The first or second argument to \`produce\` must be a function"`;

exports[`curry - proxy should check arguments 2`] = `"Invariant failed: The first or second argument to \`produce\` must be a function"`;

exports[`curry - proxy should check arguments 3`] = `"Invariant failed: The first or second argument to \`produce\` must be a function"`;

exports[`curry - proxy should check arguments 4`] = `"Invariant failed: The third argument to \`produce\` must be a function or undefined"`;
